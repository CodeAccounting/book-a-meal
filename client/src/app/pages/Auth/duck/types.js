export const AUTHENTICATING = 'AUTHENTICATING';
export const SIGNIN_SUCCESS = 'SIGNIN_SUCCESS';
export const SIGNIN_ERROR = 'SIGNIN_ERROR';
export const CLEAR_AUTH_ERROR = 'CLEAR_AUTH_ERROR';
