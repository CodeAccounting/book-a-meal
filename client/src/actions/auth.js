import instance from '../config/axios';
import { errorHandler } from '../utils';
import { auth as authAPI } from '../utils/api/auth';
import {
  AUTHENTICATED,
  UNAUTHENTICATED,
  AUTHENTICATION_ERROR,
  AUTHENTICATING,
  CLEAR_AUTH_ERROR,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_ERROR,
} from './actionTypes';

export const authenticating = () => ({
  type: AUTHENTICATING
});

export const signinSuccess = data => ({
  type: SIGNIN_SUCCESS,
  payload: data
});

export const signinFailure = error => ({
  type: SIGNIN_ERROR,
  payload: error
});

export const signupSuccess = data => ({
  type: SIGNUP_SUCCESS,
  payload: data
});

export const signupFailure = error => ({
  type: SIGNUP_ERROR,
  payload: error
});

export const authenticationSuccess = data => ({
  type: AUTHENTICATED,
  payload: data
});

export const authenticationFailure = error => ({
  type: AUTHENTICATION_ERROR,
  payload: error
});

export const clearAuthError = () => ({
  type: CLEAR_AUTH_ERROR
});

export const resetUser = () => ({
  type: UNAUTHENTICATED
});

/**
 * Authenticates User
 * @function auth
 * @param {string} type
 * @param {object} user
 * @param {func} dispatch
 * @returns {void}
 */
export const auth = type => user => async (dispatch) => {
  try {
    dispatch(authenticating());

    const response = await authAPI(type)(user);

    localStorage.setItem('jwtToken', response.data.token);

    const dispatchType = type === 'signin' ? signinSuccess : signupSuccess;

    dispatch(dispatchType(response.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    const dispatchType = type === 'signin' ? signinFailure : signupFailure;

    dispatch(dispatchType(errorResponse.response));
  }
};

/**
 * Refreshe User Token
 * @function authenticateUser
 * @param {func} dispatch
 * @returns {void}
 */
export const authenticateUser = () => async (dispatch) => {
  try {
    dispatch(authenticating());

    const res = await instance.get('/auth/refresh_token');

    localStorage.setItem('jwtToken', res.data.token);

    dispatch(authenticationSuccess(res.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    localStorage.removeItem('jwtToken');

    dispatch(authenticationFailure(errorResponse.response));
  }
};

/**
 * Logs user out
 * @function logout
 * @param {func} dispatch
 * @returns {void}
 */
export const logout = () => (dispatch) => {
  localStorage.removeItem('jwtToken');
  dispatch(resetUser());
};
