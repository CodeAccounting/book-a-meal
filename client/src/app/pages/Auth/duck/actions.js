import {
  AUTHENTICATING,
  CLEAR_AUTH_ERROR,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
} from './types';

const authenticating = () => ({
  type: AUTHENTICATING
});

const signinSuccess = data => ({
  type: SIGNIN_SUCCESS,
  payload: data
});

const signinFailure = error => ({
  type: SIGNIN_ERROR,
  payload: error
});

const clearAuthError = () => ({
  type: CLEAR_AUTH_ERROR
});

export default {
  signinSuccess,
  signinFailure,
  authenticating,
  clearAuthError,
};
