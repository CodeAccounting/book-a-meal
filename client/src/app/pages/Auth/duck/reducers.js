import {
  AUTHENTICATING,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
  CLEAR_AUTH_ERROR
} from './types';

const initialState = {
  isAuthenticated: !!localStorage.getItem('jwtToken'),
  error: null,
  user: {},
  loading: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case AUTHENTICATING:
      return { ...state, loading: true };
    case SIGNIN_SUCCESS:
      return {
        ...state,
        isAuthenticated: true,
        user: action.payload,
        error: null,
        loading: false
      };
    case CLEAR_AUTH_ERROR:
      return {
        ...state,
        isAuthenticated: false,
        error: null,
        loading: false
      };
    case SIGNIN_ERROR:
      return {
        ...state,
        isAuthenticated: false,
        user: {},
        error: action.payload,
        loading: false
      };
    default:
      return state;
  }
};
