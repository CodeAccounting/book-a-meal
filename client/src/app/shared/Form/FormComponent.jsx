import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import SigninForm from './SigninForm';
import CustomerSignupForm from './CustomerSignupForm';
import CatererSignupForm from './CatererSignupForm';
import MiniPreLoader from '../Preloader/MiniPreloader';
import { formHelpers, getTouchedFields, formErrorCount } from '../../../helpers';
import { arrayToObject } from '../../../utils';
import { syncValidate, validateRequiredFields } from '../../../helpers/validations';

/**
 * @exports
 * @class FormComponent
 * @extends Component
 * @returns {component} Form
 */
class FormComponent extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
    type: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      btnText: PropTypes.string,
      extra: PropTypes.element
    }).isRequired
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof Form
   * @param {object} props - props
   * @returns {nothing} Return nothing
   */
  constructor(props) {
    super(props);

    const { type } = props;
    const { formFields } = formHelpers;
    const fields = formFields[type];
    const values = arrayToObject(fields, '');

    if (type === 'catererSignup') values.role = 'caterer';
    if (type === 'customerSignup') values.role = 'customer';

    this.state = {
      type,
      values,
      touched: arrayToObject(fields, false),
      error: arrayToObject(fields, null),
      pristine: true,
      formValid: false,
      asyncValidating: false
    };
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  componentWillMount = () => {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  clearFormErrors = () => {
    const { type } = this.state;
    const { clearFormError } = formHelpers;

    this.props.dispatch(clearFormError[type]);
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange = (event) => {
    const { target } = event;
    const { value, name } = target;

    this.setState(prevState => ({
      values: { ...prevState.values, [name]: value },
      touched: { ...prevState.touched, [name]: true },
      pristine: false,
    }), this.validateForm);
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleFocus = (event) => {
    const { target } = event;
    const { name } = target;

    this.setState(prevState => ({
      error: { ...prevState.error, [name]: null }
    }), () => this.clearFormErrors());
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur = (event) => {
    const { target } = event;
    const { name } = target;

    this.setState(prevState => ({
      touched: { ...prevState.touched, [name]: true },
    }), () => { this.validateField(name); });
  }

  /**
   * @memberof Form
   * @param {string} name
   * @param {string} value
   * @returns {nothing} Returns nothing
   */
  validateField = (name) => {
    const { type } = this.state;
    const error = syncValidate(type)(name, this.state.values);
    const errorValue = error || null;

    this.setState(prevState => ({
      error: { ...prevState.error, [name]: errorValue }
    }), this.validateForm);
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  validateForm = () => {
    const formErrorArrayLength = formErrorCount(this.state.error);
    const touchedFields = getTouchedFields(this.state.touched);
    const { requiredFormFields } = formHelpers;
    const { type } = this.props;
    const requiredFields = requiredFormFields[type];

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength &&
      validateRequiredFields(touchedFields, requiredFields, this.state.values)) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  submitter = () => {
    const { type } = this.state;
    const { formSubmitMapper } = formHelpers;
    const { values } = this.state;

    return formSubmitMapper[type](values);
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit = (event) => {
    event.preventDefault();

    return this.props.dispatch(this.submitter());
  }

  /**
   * @memberof Form
   * @returns {JSX} Form
   */
  renderForm = () => {
    const { type } = this.props;
    const formState = { ...this.state };

    const handlers = {
      handleChange: this.handleChange,
      handleBlur: this.handleBlur,
      handleFocus: this.handleFocus,
      handleSubmit: this.handleSubmit
    };

    switch (type) {
      case 'customerSignup':
        return <CustomerSignupForm type={type} state={formState} handlers={handlers} />;
      case 'catererSignup':
        return <CatererSignupForm type={type} state={formState} handlers={handlers} />;
      default:
        return <SigninForm type={type} state={formState} handlers={handlers} />;
    }
  }

  /**
   * @memberof Form
   * @returns {component} Form
   */
  render() {
    const { pristine, formValid } = this.state;
    const { submitting, submitError, meta: { btnText, extra } } = this.props;
    const requiredTextArray = ['catererSignup', 'customerSignup'];

    return (
      <div>
        {submitting &&
          <div className="modal-preloader text-center"><MiniPreLoader /></div>}
        {!submitting &&
          <Fragment>
            {requiredTextArray.includes(this.props.type) &&
            <p className="text-muted mx-auto text-center">
                Fields marked
              <span className="danger"> *</span> are required
            </p>}
            <form onSubmit={this.handleSubmit}>
              {submitError && (
              <p className="danger text-center mb-0">{submitError}</p>
              )}
              {this.renderForm()}
              <button
                className="btn btn-pri btn-block"
                disabled={!formValid || !!pristine || !!submitting}
              >
                {btnText}
              </button>
            </form>
            {extra && extra}
          </Fragment>}
      </div>);
  }
}

export default FormComponent;
