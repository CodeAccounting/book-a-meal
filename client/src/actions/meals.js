import { toastr } from 'react-redux-toastr';
import instance from '../config/axios';
import errorHandler from '../utils/errorHandler';
import { setFetching, unsetFetching } from '../actions/isFetching';
import { toggleModal } from '../actions/ui';
import {
  RECEIVE_MEALS_SUCCESS, RECEIVE_MEALS_FAILURE, SET_MEAL_WORKING,
  UNSET_MEAL_WORKING, CLEAR_MEAL_ERROR, ADD_MEAL_SUCCESS, ADD_MEAL_FAILURE,
  EDIT_MEAL_SUCCESS, EDIT_MEAL_FAILURE, DELETE_MEAL_SUCCESS, DELETE_MEAL_FAILURE,
  RECEIVE_MORE_MEALS_SUCCESS
} from './actionTypes';

export const fetchMealsSuccess = (type, payload) => ({
  type,
  payload
});

export const fetchMealsFailure = payload => ({
  type: RECEIVE_MEALS_FAILURE,
  payload
});

export const setMealWorking = () => ({
  type: SET_MEAL_WORKING
});

export const unsetMealWorking = () => ({
  type: UNSET_MEAL_WORKING
});

export const addMealSuccess = payload => ({
  type: ADD_MEAL_SUCCESS,
  payload
});

export const addMealFailure = payload => ({
  type: ADD_MEAL_FAILURE,
  payload
});

export const editMealSuccess = payload => ({
  type: EDIT_MEAL_SUCCESS,
  payload
});

export const editMealFailure = payload => ({
  type: EDIT_MEAL_FAILURE,
  payload
});

export const deleteMealSuccess = payload => ({
  type: DELETE_MEAL_SUCCESS,
  payload
});

export const deleteMealFailure = payload => ({
  type: DELETE_MEAL_FAILURE,
  payload
});

export const clearMealError = () => ({
  type: CLEAR_MEAL_ERROR
});

/**
 * Fetchs Meals for Caterer
 * @function fetchMeals
 * @param {object} metadata
 * @param {func} dispatch
 * @returns {void}
 */
export const fetchMeals = metadata => async (dispatch) => {
  const url = (metadata && metadata.next) || '/meals';

  try {
    if (!metadata) dispatch(setFetching());

    const response = await instance.get(url);

    const type = metadata ? RECEIVE_MORE_MEALS_SUCCESS : RECEIVE_MEALS_SUCCESS;

    dispatch(fetchMealsSuccess(type, response.data));
    dispatch(unsetFetching());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(fetchMealsFailure(errorResponse.response));
    dispatch(unsetFetching());
  }
};

/**
 * Adds a Meal for Caterer
 * @function addMeal
 * @param {object} meal
 * @param {func} dispatch
 * @returns {void}
 */
export const addMeal = meal => async (dispatch) => {
  try {
    dispatch(setMealWorking());

    const response = await instance.post('/meals', meal);

    dispatch(addMealSuccess(response.data));
    toastr.success('Meal Added Successfully');
    dispatch(unsetMealWorking());
    dispatch(toggleModal());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(addMealFailure(errorResponse.response));
    dispatch(unsetMealWorking());
  }
};

/**
 * Edits a Meal
 * @function editMeal
 * @param {string} id
 * @param {object} updatedMeal
 * @param {func} dispatch
 * @returns {void}
 */
export const editMeal = (id, updatedMeal) => async (dispatch) => {
  try {
    dispatch(setMealWorking());

    const response = await instance.put(`/meals/${id}`, updatedMeal);

    dispatch(editMealSuccess(response.data));
    toastr.success('Meal Modified Successfully');
    dispatch(unsetMealWorking());
    dispatch(toggleModal());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(editMealFailure(errorResponse.response));
    dispatch(unsetMealWorking());
  }
};

/**
 * Deletes a Meal
 * @function deleteMeal
 * @param {id} id
 * @param {func} dispatch
 * @returns {void}
 */
export const deleteMeal = id => async (dispatch) => {
  try {
    dispatch(setMealWorking());

    await instance.delete(`/meals/${id}`);

    dispatch(deleteMealSuccess(id));
    toastr.success('Meal Deleted Successfully');
    dispatch(unsetMealWorking());
    dispatch(toggleModal());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(deleteMealFailure(errorResponse.response));
    dispatch(unsetMealWorking());
  }
};
