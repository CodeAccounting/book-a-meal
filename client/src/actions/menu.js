import moment from 'moment';
import { toastr } from 'react-redux-toastr';
import instance from '../config/axios';
import errorHandler from '../utils/errorHandler';
import { setFetching, unsetFetching } from '../actions/isFetching';
import { toggleModal } from '../actions/ui';
import {
  RECEIVE_MENU_SUCCESS, RECEIVE_MENU_FAILURE, SET_CURRENT_DAY, SET_MENU_WORKING,
  UNSET_MENU_WORKING, CLEAR_MENU_ERROR, ADD_MENU_SUCCESS, ADD_MENU_FAILURE,
  EDIT_MENU_SUCCESS, EDIT_MENU_FAILURE, RECEIVE_MORE_MENU_SUCCESS,
} from './actionTypes';

export const fetchMenuSuccess = (type, payload) => ({
  type,
  payload
});

export const setMenuWorking = () => ({
  type: SET_MENU_WORKING
});

export const unsetMenuWorking = () => ({
  type: UNSET_MENU_WORKING
});

export const clearMenuError = () => ({
  type: CLEAR_MENU_ERROR
});

export const fetchMenuFailure = payload => ({
  type: RECEIVE_MENU_FAILURE,
  payload
});

export const setCurrentDay = payload => ({
  type: SET_CURRENT_DAY,
  payload
});

export const addMenuSuccess = payload => ({
  type: ADD_MENU_SUCCESS,
  payload
});

export const addMenuFailure = payload => ({
  type: ADD_MENU_FAILURE,
  payload
});

export const editMenuSuccess = payload => ({
  type: EDIT_MENU_SUCCESS,
  payload
});

export const editMenuFailure = payload => ({
  type: EDIT_MENU_FAILURE,
  payload
});

/**
 * Fetchs Menu
 * @function fetchMenu
 * @param {string} date
 * @param {object} metadata for pagination
 * @param {func} dispatch
 * @returns {void}
 */
export const fetchMenu = (date, metadata) => async (dispatch) => {
  const menuDate = date || moment().format('YYYY-MM-DD');
  const url = (metadata && metadata.next) || `/menu?date=${menuDate}`;

  try {
    if (!metadata) dispatch(setFetching());

    const response = await instance.get(url);

    const type = metadata ? RECEIVE_MORE_MENU_SUCCESS : RECEIVE_MENU_SUCCESS;

    dispatch(fetchMenuSuccess(type, response.data));

    dispatch(unsetFetching());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(fetchMenuFailure(errorResponse.response));
    dispatch(unsetFetching());
  }
};

/**
 * Adds a Menu for Caterer
 * @function addMenu
 * @param {object} menu
 * @param {func} dispatch
 * @returns {void}
 */
export const addMenu = menu => async (dispatch) => {
  try {
    dispatch(setMenuWorking());

    const response = await instance.post('/menu', menu);

    dispatch(addMenuSuccess(response.data));
    toastr.success('Menu Set Successfully');
    dispatch(unsetMenuWorking());
    dispatch(toggleModal());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(addMenuFailure(errorResponse.response));
    dispatch(unsetMenuWorking());
  }
};

/**
 * Edits a Menu
 * @function editMenu
 * @param {string} id
 * @param {object} menu
 * @param {func} dispatch
 * @returns {void}
 */
export const editMenu = (id, menu) => async (dispatch) => {
  try {
    dispatch(setMenuWorking());

    const response = await instance.put(`/menu/${id}`, menu);

    dispatch(editMenuSuccess(response.data));
    toastr.success('Menu Set Successfully');
    dispatch(unsetMenuWorking());
    dispatch(toggleModal());
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(editMenuFailure(errorResponse.response));
    dispatch(unsetMenuWorking());
  }
};
