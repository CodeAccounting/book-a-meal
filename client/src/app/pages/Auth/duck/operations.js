import { errorHandler } from '../../../../utils';
import { auth as authAPI } from '../../../../services/api/auth';
import {
  signinSuccess,
  signinFailure,
  authenticating,
  clearAuthError,
} from './actions';

const auth = type => user => async (dispatch) => {
  try {
    dispatch(authenticating());

    const response = await authAPI(type)(user);

    localStorage.setItem('jwtToken', response.data.token);

    dispatch(signinSuccess(response.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    dispatch(signinFailure(errorResponse.response));
  }
};

export default {
  auth,
  signinSuccess,
  signinFailure,
  authenticating,
  clearAuthError,
};
