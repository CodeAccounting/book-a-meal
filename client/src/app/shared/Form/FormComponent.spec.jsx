import React from 'react';
import FormComponent from './FormComponent';
import { clearAuthError } from '../../pages/Auth/duck/actions';
import { mainFormSetup as setup } from '../../../../tests/setup/formSetup';

const state = {
  type: 'signin',
  values: {
    email: '',
    password: ''
  },
  touched: { email: false, password: false },
  error: { email: null, password: null },
  pristine: true,
  formValid: false,
  asyncValidating: false
};

const signinMeta = {
  title: 'Sign In to Your Account',
  btnText: 'SIGN IN',
  extra: <p>something</p>
};

describe('Form', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { shallowRoot } = setup('signin', signinMeta);

    expect(toJson(shallowRoot)).toMatchSnapshot();
  });


  it('disables submit button when form is clean', () => {
    const { shallowRoot } = setup('signin', signinMeta);

    expect(shallowRoot.find('button[disabled=true]')).toBeTruthy();
  });

  it('disables submit button when submitting form', () => {
    const { props, dispatchMock } = setup('signin', signinMeta);
    const newProps = {
      ...props,
      submitting: true
    };
    const shallowRoot = shallow(<FormComponent {...newProps} dispatch={dispatchMock} />);

    expect(toJson(shallowRoot)).toMatchSnapshot();
    expect(shallowRoot.find('button[disabled=true]')).toBeTruthy();
  });

  it('shows error alert and disables submit button when there\'s a submit error', () => {
    const { props, dispatchMock } = setup('signin', signinMeta);
    const newProps = {
      ...props,
      submitError: 'Username/Password do not match'
    };

    const shallowRoot = shallow(<FormComponent {...newProps} dispatch={dispatchMock} />);

    expect(toJson(shallowRoot)).toMatchSnapshot();
    expect(shallowRoot.find('p.danger')).toBeTruthy();
    expect(shallowRoot.find('button[disabled=true]')).toBeTruthy();
  });

  describe('test for right input', () => {
    it('calls handleChange and handleBlur on input change and blur for email field', (done) => {
      const { mountRoot, dispatchMock } = setup('signin', signinMeta);
      const wrapper = mountRoot.find(FormComponent);

      const changeState = {
        ...state,
        values: { ...state.values, email: 'iverenshaguy@gmail.com' },
        touched: { ...state.touched, email: true },
        error: { email: null, password: null },
        pristine: false
      };

      const blurState = {
        ...changeState,
        asyncValidating: false
      };

      const event = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };

      expect(toJson(wrapper)).toMatchSnapshot();

      wrapper.find('input[name="email"]').simulate('focus');
      expect(dispatchMock).toHaveBeenCalledWith(clearAuthError());

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="email"]').simulate('blur', event);

      setTimeout(() => {
        expect(wrapper.instance().state).toEqual(blurState);
        done();
      }, 600);
    });

    it('submits valid form', () => {
      const { mountRoot, dispatchMock } = setup('signin', signinMeta);
      const wrapper = mountRoot.find(FormComponent);

      const newState = {
        ...state,
        values: { email: 'iverenshaguy@gmail.com', password: 'iverenshaguy' },
        touched: { email: true, password: true },
        error: { email: null, password: null },
        pristine: false,
        formValid: true
      };

      const emailEvent = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };
      const passwordEvent = { target: { name: 'password', value: 'iverenshaguy' } };

      wrapper.find('input[name="email"]').simulate('focus');
      wrapper.find('input[name="email"]').simulate('change', emailEvent);
      wrapper.find('input[name="password"]').simulate('focus');
      wrapper.find('input[name="password"]').simulate('change', passwordEvent);

      expect(wrapper.instance().state).toEqual(newState);

      wrapper.find('form').simulate('submit', { preventDefault() { } });
      expect(dispatchMock).toHaveBeenCalled();

      mountRoot.unmount();
    });
  });

  describe('test for wrong input', () => {
    it('sync validates field and form on input change and blur', () => {
      const { mountRoot } = setup('signin', signinMeta);
      const wrapper = mountRoot.find(FormComponent);


      const changeState = {
        ...state,
        values: { ...state.values, email: 'emiolaolasanmi' },
        touched: { ...state.touched, email: true },
        error: { email: 'Invalid email address!', password: null },
        pristine: false
      };

      const blurState = {
        ...changeState,
        asyncValidating: false
      };

      const event = { target: { name: 'email', value: 'emiolaolasanmi' } };

      wrapper.find('input[name="email"]').simulate('focus');
      wrapper.find('input[name="email"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(({
        ...changeState, error: { email: 'Required!', password: null }, pristine: true, values: { ...state.values }
      }));

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual({ ...changeState, error: { ...changeState.error, email: 'Required!' } });

      wrapper.find('input[name="email"]').simulate('blur', event);
      expect(wrapper.instance().state).toEqual(blurState);

      wrapper.find('input[name="password"]').simulate('focus');
      wrapper.find('input[name="password"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(({
        ...changeState,
        error: { ...changeState.error, password: 'Required!' },
        touched: { email: true, password: true },
        values: { email: 'emiolaolasanmi', password: '' }
      }));

      mountRoot.unmount();
    });
  });
});
