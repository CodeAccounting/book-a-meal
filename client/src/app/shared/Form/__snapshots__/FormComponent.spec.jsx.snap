// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form disables submit button when submitting form 1`] = `
<div>
  <div
    className="modal-preloader text-center"
  >
    <MiniPreLoader />
  </div>
</div>
`;

exports[`Form renders correctly 1`] = `
<div>
  <React.Fragment>
    <form
      onSubmit={[Function]}
    >
      <SigninForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "email": null,
              "password": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "email": false,
              "password": false,
            },
            "values": Object {
              "email": "",
              "password": "",
            },
          }
        }
        type="signin"
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form shows error alert and disables submit button when there's a submit error 1`] = `
<div>
  <React.Fragment>
    <form
      onSubmit={[Function]}
    >
      <p
        className="danger text-center mb-0"
      >
        Username/Password do not match
      </p>
      <SigninForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "email": null,
              "password": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "email": false,
              "password": false,
            },
            "values": Object {
              "email": "",
              "password": "",
            },
          }
        }
        type="signin"
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form test for right input calls handleChange and handleBlur on input change and blur for email field 1`] = `
<FormComponent
  dispatch={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
        Array [
          Object {
            "type": "CLEAR_AUTH_ERROR",
          },
        ],
      ],
    }
  }
  meta={
    Object {
      "btnText": "SIGN IN",
      "extra": <p>
        something
      </p>,
      "title": "Sign In to Your Account",
    }
  }
  submitError={null}
  submitting={false}
  type="signin"
>
  <div>
    <form
      onSubmit={[Function]}
    >
      <SigninForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "email": null,
              "password": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "email": false,
              "password": false,
            },
            "values": Object {
              "email": "",
              "password": "",
            },
          }
        }
        type="signin"
      >
        <RenderInput
          handleBlur={[Function]}
          handleChange={[Function]}
          handleFocus={[Function]}
          id="email"
          label="Email Address"
          meta={
            Object {
              "asyncValidating": false,
              "error": null,
              "touched": false,
            }
          }
          name="email"
          placeholder=""
          required={true}
          type="email"
          value=""
        >
          <div
            className="form-input"
          >
            <label
              htmlFor="email"
            >
              Email Address
              <span
                className="danger"
              >
                *
              </span>
            </label>
            <input
              className=""
              id="email"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="email"
              value=""
            />
          </div>
        </RenderInput>
        <RenderInput
          handleBlur={[Function]}
          handleChange={[Function]}
          handleFocus={[Function]}
          id="password"
          label="Password"
          meta={
            Object {
              "error": null,
              "touched": false,
            }
          }
          name="password"
          placeholder=""
          required={true}
          type="password"
          value=""
        >
          <div
            className="form-input"
          >
            <label
              htmlFor="password"
            >
              Password
              <span
                className="danger"
              >
                *
              </span>
            </label>
            <input
              className=""
              id="password"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="password"
              value=""
            />
          </div>
        </RenderInput>
      </SigninForm>
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </div>
</FormComponent>
`;
